vault/identity_store_oidc.go

```diff
@@ -1346,35 +1472,79 @@ func (i *IdentityStore) oidcKeyRotation(ctx context.Context, s logical.Storage)
 
 // oidcPeriodFunc is invoked by the backend's periodFunc and runs regular key
 // rotations and expiration actions.
-func (i *IdentityStore) oidcPeriodicFunc(ctx context.Context, s logical.Storage) {
-	nextRun := time.Time{}
+func (i *IdentityStore) oidcPeriodicFunc(ctx context.Context) {
+	var nextRun time.Time
+	now := time.Now()
 
-	if v, ok := i.oidcCache.Get("nextRun"); ok {
+	nsPaths := i.listNamespacePaths(ctx)
+
+	if v, ok := i.oidcCache.Get(nil, "nextRun"); ok {
 		nextRun = v.(time.Time)
 	}
 
 	// The condition here is for performance, not precise timing. The actions can
 	// be run at any time safely, but there is no need to invoke them (which
 	// might be somewhat expensive if there are many roles/keys) if we're not
 	// past any rotation/expiration TTLs.
-	if time.Now().After(nextRun) {
-		nextRotation, err := i.oidcKeyRotation(ctx, s)
-		if err != nil {
-			i.Logger().Warn("error rotating OIDC keys", "err", err)
-		}
+	if now.After(nextRun) {
+		// Initialize to a fairly distant next run time. This will be brought in
+		// based on key rotation times.
+		nextRun = now.Add(24 * time.Hour)
 
-		nextExpiration, err := i.expireOIDCPublicKeys(ctx, s)
-		if err != nil {
-			i.Logger().Warn("error expiring OIDC public keys", "err", err)
-		}
+		for _, nsPath := range nsPaths {
+			s := i.core.router.MatchingStorageByAPIPath(ctx, nsPath+"identity/oidc")
 
-		i.oidcCache.Flush()
+			if s == nil {
+				continue
+			}
+
+			nextRotation, err := i.oidcKeyRotation(ctx, s)
+			if err != nil {
+				i.Logger().Warn("error rotating OIDC keys", "err", err)
+			}
+
+			nextExpiration, err := i.expireOIDCPublicKeys(ctx, s)
+			if err != nil {
+				i.Logger().Warn("error expiring OIDC public keys", "err", err)
+			}
+
+			i.oidcCache.Flush(nil)
+
+			// re-run at the soonest expiration or rotation time
+			if nextRotation.Before(nextRun) {
+				nextRun = nextRotation
+			}
 
-		// re-run at the soonest expiration or rotation time
-		if nextRotation.Before(nextExpiration) {
-			i.oidcCache.SetDefault("nextRun", nextRotation)
-		} else {
-			i.oidcCache.SetDefault("nextRun", nextExpiration)
+			if nextExpiration.Before(nextRun) {
+				nextRun = nextExpiration
+			}
 		}
+		i.oidcCache.SetDefault(nil, "nextRun", nextRun)
+	}
+}
+
+func newOIDCCache() *oidcCache {
+	return &oidcCache{
+		c: cache.New(cache.NoExpiration, cache.NoExpiration),
+	}
+}
+
+func (c *oidcCache) nskey(ns *namespace.Namespace, key string) string {
+	if ns == nil {
+		ns = &namespace.Namespace{ID: "__root__"}
```

Comment:
Why not use https://github.com/hashicorp/vault/blob/master/helper/namespace/namespace.go#L17?

Reply:
I actually changed it (in a yet un-pushed minor PR), to "", to distinguish between `ns==nil` and the root namespace.

Reply:
Is there a case where `ns` would be `nil`?

Reply:
The idea (which is admittedly new) was to be able to store values that are not associated with any namespace... i.e. pass in `nil` explicitly if needed, vs. the namespace from a context. This arose because of things like the periodicFunc that's doing background ops across all namespaces, and isn't attached to any one in particular.

Reply:
I think it might be better to use a const, similar to root, that can represent "None". This would make operations that rely on this value explicitly check against it and guard against the case where it might have been accidentally left out. I believe that there are checks in Vault around a context not having a namespace so you'd have to work around those if you went with using empty/nil namespace values.

Reply:
I like that idea

Reply:
Converted to a dedicated, non-nil namespace.